File: smoke_tests/typescript/xpath_injection.ts
// XPath Injection Test Cases

// Test 1: XPath query with unsanitized user input
function findUserByName(xmlDoc: any, username: string): any {
    // VULNERABLE: username could contain XPath injection like ' or '1'='1
    const xpath = `//users/user[username/text()='${username}']`;
    return xmlDoc.evaluate(xpath);
}

// Test 2: Authentication using XPath
function authenticateWithXPath(xmlDoc: any, user: string, pass: string): boolean {
    // VULNERABLE: user could be "admin' or '1'='1" to bypass authentication
    const xpath = `//user[username='${user}' and password='${pass}']`;
    const result = xmlDoc.evaluate(xpath);
    return result.length > 0;
}

// Test 3: XPath query with multiple user inputs
function searchProducts(xmlDoc: any, category: string, minPrice: string): any {
    // VULNERABLE: Both category and minPrice are unsanitized
    const xpath = `//products/product[category='${category}' and price>=${minPrice}]`;
    return xmlDoc.evaluate(xpath);
}

// Test 4: XPath contains function with user input
function searchByPartialMatch(xmlDoc: any, searchTerm: string): any {
    // VULNERABLE: searchTerm could break out of the contains function
    const xpath = `//item[contains(name, '${searchTerm}')]`;
    return xmlDoc.evaluate(xpath);
}

// Test 5: Complex XPath with user-controlled predicate
function customXPathQuery(xmlDoc: any, field: string, value: string): any {
    // VULNERABLE: field and value allow arbitrary XPath construction
    const xpath = `//${field}[text()='${value}']`;
    return xmlDoc.evaluate(xpath);
}

File: smoke_tests/rust/xpath_injection.rs
// XPath Injection Test Cases

// Test 1: XPath query with unsanitized user input
fn find_user_by_name(username: &str) -> String {
    // VULNERABLE: username could contain XPath injection like ' or '1'='1
    format!("//users/user[username/text()='{}']", username)
}

// Test 2: Authentication using XPath
fn authenticate_xpath(user: &str, pass: &str) -> String {
    // VULNERABLE: user could be "admin' or '1'='1" to bypass authentication
    format!("//user[username='{}' and password='{}']", user, pass)
}

// Test 3: XPath query with multiple user inputs
fn search_products(category: &str, min_price: &str) -> String {
    // VULNERABLE: Both category and min_price are unsanitized
    format!("//products/product[category='{}' and price>={}]", category, min_price)
}

// Test 4: XPath contains function with user input
fn search_by_partial_match(search_term: &str) -> String {
    // VULNERABLE: search_term could break out of the contains function
    format!("//item[contains(name, '{}')]", search_term)
}

// Test 5: Complex XPath with user-controlled predicate
fn custom_xpath_query(field: &str, value: &str) -> String {
    // VULNERABLE: field and value allow arbitrary XPath construction
    format!("//{}[text()='{}']", field, value)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_xpath_injection() {
        let query = find_user_by_name("admin");
        assert!(query.contains("admin"));
    }
}
